keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+	// that will be used by the
+	// driver when allocating a resource for the claim.
+	//
+	// The object must be in the same namespace as the ResourceClaim.
+	// An empty namespace field defaults to that namespace.
+	// ApiVersion and name must be set. Kind and/or resource must be set.
+	// Usually, kind is sufficient because there will be a 1:1 mapping
+	// to resource. If UID is set, then only that specific object
+	// may be used.
+	Parameters v1.ObjectReference
+
+	// Allocation can start immediately or when a Pod wants to use the
+	// resource. Waiting for a Pod is the default.
+	AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a ResourceClaim gets allocated immediately
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod (AllocationModeDelayed). Other modes
+// might get added in the future.
+type AllocationMode string
+
+const (
+	// When a ResourceClaim has AllocationModeDelayed, allocation is
+	// delayed until a Pod gets scheduled that needs the ResourceClaim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeDelayed AllocationMode = "Delayed"
+
+	// When a ResourceClaim has AllocationModeImmediate, allocation starts
+	// as soon as the ResourceClaim gets created. This is done without
+	// considering the needs of Pods that will use the ResourceClaim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ResourceClaimStatus struct {
+	// DriverName is a copy of the driver name from the ResourceClass at
+	// the time when allocation started. It's necessary to support
+	// deallocation when the class gets deleted before a claim.
+	DriverName string
+
+	// Scheduling contains information that is only relevant while the
+	// scheduler and the resource driver are in the process of selecting a
+	// node for a Pod and the allocation mode is AllocationModeDelayed. The
+	// resource driver should unset this when it has successfully allocated
+	// the resource.
+	Scheduling SchedulingStatus
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. Nil indicates that the resource is not
+	// allocated.
+	Allocation *AllocationResult
+
+	// Deallocate may be set to true to request freeing a resource as soon
+	// as it is unused. The scheduler uses this when it finds that freeing
+	// the resource and reallocating it elsewhere might unblock a pod.
+	//
+	// The resource driver checks this fields and resets it to false
+	// together with clearing the Allocation field. It also sets it
+	// to false when the resource is not allocated.
+	Deallocate bool
+
+	// ReservedFor indicates which entities are currently allowed to use
```

Comment:
Is this bounded? E.g. could I, theoretically, use a resource on 5000 pods at the same time? We can't have unbounded lists.

Reply:
It simply can't be unbounded.  We don't have an analog in volumes - we rely on the underlying implementation to prevent problems.

As a counter-point, we DO have large, unbounded lists in Node status (volumes and images) and it has been a problem.  EndpointSlice exists BECAUSE of this problem.

So, if we need a way to list which nodes a device is present on, we need a different way to do it.  We could think about a `ResourceClaimNode` type which maps a `ResourceClaim` to a single `Node` (or small set of).  I am sure there must be other ways.

Reply:
> It simply can't be unbounded. 

But some implementation-defined limit (say, the magic 16 or 32) would be okay? It may be good enough in practice, or at least as a starting point.

We can also limit it to 1 and make this single value, but then we give up one of the goals of the KEP and need to change the API if we want to bring it back. This may be okay. I don't know at this point.

We need a way to determine whether a pod is about to start that uses a certain claim, so removing this completely is problematic.

Reply:
If you want to go into alpha with a fixed size list, I won't block it, but given the desire to have multi-user resources, I am suspicious of a fixed limit.

Reply:
I'm also not happy about it, but I don't see any easy solution. Let's defer for now:
```
	// The maximum size is 32. This is an artificial limit to prevent
	// a completely unbounded field in the API.
	ReservedFor []metav1.OwnerReference

	<<[UNRESOLVED pohly]>>
	We will have to discuss use cases and real resource drivers that
	support sharing before deciding on a) which limit is useful and
	b) whether we need a different API that supports an arbitrary
	number of users.

	Any solution that handles reservations differently will have to
	be very careful about race conditions.
	<<[/UNRESOLVED]>>
```

I've not added it as beta criteria because I consider an UNRESOLVED comment in the API as automatic blocker for that.